src/translate/pddl_parser/parsing_functions.py

```diff
@@ -188,121 +314,229 @@ def add_effect(tmp_effect, result):
+                parameters_tag_opt = next(iterator)
+                if parameters_tag_opt == ":parameters":
+                    parameters_list = next(iterator)
+                    context.check_list(parameters_list, block_name="Parameters")
+                    parameters = parse_typed_list(context, parameters_list,
+                                                  only_variables=True)
+                    precondition_tag_opt = next(iterator)
+                else:
+                    parameters = []
+                    precondition_tag_opt = parameters_tag_opt
+            with context.layer("Parsing precondition"):
+                if precondition_tag_opt == ":precondition":
+                    precondition_list = next(iterator)
+                    context.check_list(precondition_list, block_name="Precondition")
+                    if not precondition_list:
+                        # Note that :precondition () is allowed in PDDL.
+                        precondition = pddl.Conjunction([])
+                    else:
+                        precondition = parse_condition(
+                            context, precondition_list, type_dict, predicate_dict)
+                    effect_tag = next(iterator)
+                else:
+                    precondition = pddl.Conjunction([])
+                    effect_tag = precondition_tag_opt
+            with context.layer("Parsing effect"):
+                context.check_string(effect_tag, value=":effect", item_name="Effect tag",
+                                     finalmsg=f"Expecting '{syntax}'.")
+                effect_list = next(iterator)
+                context.check_list(effect_list, block_name="Effect")
+                eff = []
+                if effect_list:
+                    try:
+                        cost = parse_effects(
+                            context, effect_list, eff, type_dict, predicate_dict)
+                    except ValueError as e:
+                        context.error("Error in Action %s\nReason: %s." % (name, e))
+        except StopIteration:
+            context.error(f"Missing fields. Expecting {syntax}.")
+
     if eff:
         return pddl.Action(name, parameters, len(parameters),
                            precondition, eff, cost)
     else:
         return None
 
 
-def parse_axiom(alist, type_dict, predicate_dict):
-    assert len(alist) == 3
-    assert alist[0] == ":derived"
-    predicate = parse_predicate(alist[1])
-    condition = parse_condition(
-        alist[2], type_dict, predicate_dict)
-    return pddl.Axiom(predicate.name, predicate.arguments,
-                      len(predicate.arguments), condition)
+def parse_axiom(context, no, alist, type_dict, predicate_dict):
+    syntax = "'(:derived PREDICATE CONDITION)'"
+    with context.layer(f"Parsing {no}. axiom"):
+        if len(alist) != 3:
+            context.error(f"Expecting {syntax}.")
+        assert alist[0] == ":derived"
```

Comment:
in other places this is done with `check_expected_block` or `check_string`. Why is it an assert here? Is this checked by the calling code?